go/cmd/exporter/exporter.go

```diff
@@ -0,0 +1,170 @@
+	vulnBucketName := flag.String("osv_vulns_bucket", os.Getenv("OSV_VULNERABILITIES_BUCKET"), "GCS bucket to read vulnerability protobufs from. Can also be set with the OSV_VULNERABILITIES_BUCKET environment variable.")
+	local := flag.Bool("local", true, "If true, writes the output to a local directory specified by -bucket instead of a GCS bucket.")
+	numWorkers := flag.Int("num_workers", 200, "The total number of concurrent workers to use for downloading from GCS and writing the output.")
+
+	flag.Parse()
+
+	logger.Info("exporter starting",
+		slog.String("bucket", *outBucketName),
+		slog.String("osv_vulns_bucket", *vulnBucketName),
+		slog.Bool("local", *local),
+		slog.Int("num_workers", *numWorkers))
+
+	if *vulnBucketName == "" {
+		logger.Fatal("OSV_VULNERABILITIES_BUCKET must be set")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	cl, err := storage.NewClient(ctx)
+	if err != nil {
+		logger.Fatal("failed to create storage client", slog.Any("err", err))
+	}
+
+	vulnBucket := cl.Bucket(*vulnBucketName)
+	var outBucket *storage.BucketHandle
+	var outPrefix string
+	if *local {
+		outPrefix = *outBucketName
+	} else {
+		outBucket = cl.Bucket(*outBucketName)
+	}
+
+	objCh := make(chan *storage.ObjectHandle)
+	resultsCh := make(chan *osvschema.Vulnerability)
+	writeCh := make(chan writeMsg)
+
+	var downloaderWg sync.WaitGroup
+	for range *numWorkers / 2 {
+		downloaderWg.Add(1)
+		go downloader(ctx, objCh, resultsCh, &downloaderWg)
+	}
+
+	var writerWg sync.WaitGroup
+	for range *numWorkers / 2 {
+		writerWg.Add(1)
+		go writer(ctx, cancel, writeCh, outBucket, outPrefix, &writerWg)
+	}
+	var routerWg sync.WaitGroup
+	routerWg.Add(1)
+	go ecosystemRouter(ctx, resultsCh, writeCh, &routerWg)
+
+	it := vulnBucket.Objects(ctx, &storage.Query{Prefix: "all/pb/"})
+	prevPrefix := ""
+MainLoop:
+	for {
+		attrs, err := it.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			logger.Fatal("failed to list objects", slog.Any("err", err))
+		}
+		// Only log when we see a new ID prefix (i.e. roughly once per data source)
+		prefix := filepath.Base(attrs.Name)
+		prefix, _, _ = strings.Cut(prefix, "-")
+		if prefix != prevPrefix {
+			logger.Info("iterating vulnerabilities", slog.String("now_at", attrs.Name))
+			prevPrefix = prefix
+		}
+		select {
+		case objCh <- vulnBucket.Object(attrs.Name):
+		case <-ctx.Done():
+			break MainLoop
+		}
+	}
+
+	close(objCh)
+	downloaderWg.Wait()
+	close(resultsCh)
+	routerWg.Wait()
+	close(writeCh)
+	writerWg.Wait()
+
+	if ctx.Err() != nil {
+		logger.Fatal("exporter cancelled")
+	}
+	logger.Info("export completed successfully")
+}
+
+func ecosystemRouter(ctx context.Context, inCh <-chan *osvschema.Vulnerability, writeCh chan<- writeMsg, wg *sync.WaitGroup) {
+	defer wg.Done()
+	logger.Info("ecosystem router starting")
+	workers := make(map[string]*ecosystemWorker)
+	var workersWg sync.WaitGroup
+	var vulnCounter int
+
+	workersWg.Add(1)
```

Comment:
nit: Maybe here and newEcosystemWorker as well the Add(1) should be inside the new function to avoid it from being forgotten (and consistently next to the `go` keyword)